lib/db_connection/connection_pool.ex

```diff
@@ -0,0 +1,428 @@
+    GenServer.start_link(__MODULE__, {mod, opts}, start_opts(opts))
+  end
+
+  @doc false
+  def child_spec(mod, opts, child_opts \\ []) do
+    Supervisor.Spec.worker(__MODULE__, [mod, opts], child_opts)
+  end
+
+  @doc false
+  def checkout(pool, opts) do
+    queue? = Keyword.get(opts, :queue, @queue)
+    now = System.monotonic_time(@time_unit)
+    timeout = abs_timeout(now, opts)
+    case checkout(pool, queue?, now, timeout) do
+      {:ok, _, _, _} = ok ->
+        ok
+      {:error, _} = error ->
+        error
+      {:exit, reason} ->
+        exit({reason, {__MODULE__, :checkout, [pool, opts]}})
+    end
+  end
+
+  @doc false
+  def checkin({pool, ref, deadline, holder}, conn, _) do
+    cancel_deadline(deadline)
+    now = System.monotonic_time(@time_unit)
+    checkin_holder(holder, pool, conn, {:checkin, ref, now})
+  end
+
+  @doc false
+  def disconnect({pool, ref, deadline, holder}, err, conn, _) do
+    cancel_deadline(deadline)
+    checkin_holder(holder, pool, conn, {:disconnect, ref, err})
+  end
+
+  @doc false
+  def stop({pool, ref, deadline, holder}, err, conn, _) do
+    cancel_deadline(deadline)
+    checkin_holder(holder, pool, conn, {:stop, ref, err})
+  end
+
+  ## Holder api
+
+  @doc false
+  def update(pool, ref, mod, state) do
+    holder = start_holder(pool, ref, mod, state)
+    now = System.monotonic_time(@time_unit)
+    checkin_holder(holder, pool, state, {:checkin, ref, now})
+    holder
+  end
+
+  ## GenServer api
+    
+  def init({mod, opts}) do
+    queue = :ets.new(__MODULE__.Queue, [:private, :ordered_set])
+    {:ok, _} = PoolSupervisor.start_pool(queue, mod, opts)
+    target = Keyword.get(opts, :queue_target, @queue_target)
+    interval = Keyword.get(opts, :queue_interval, @queue_interval)
+    idle_interval = Keyword.get(opts, :idle_interval, @idle_interval)
+    now = System.monotonic_time(@time_unit)
+    codel = %{target: target, interval: interval, delay: 0, slow: false,
+              next: now, poll: nil, idle_interval: idle_interval, idle: nil}
+    codel = start_idle(now, now, start_poll(now, now, codel))
+    {:ok, {:busy, queue, codel}}
+  end
+
+  def handle_info({:db_connection, from, {:checkout, now, queue?}}, {:busy, queue, _} = busy) do
+    case queue? do
+      true ->
+        :ets.insert(queue, {{now, System.unique_integer(), from}})
+        {:noreply, busy}
+      false ->
+        message = "connection not available and queuing is disabled"
+        err = DBConnection.ConnectionError.exception(message)
+        GenServer.reply(from, {:error, err})
+        {:noreply, busy}
+    end
+  end
+
+  def handle_info({:db_connection, from, {:checkout, _now, _queue?}} = checkout, ready) do
+    {:ready, queue, _codel} = ready 
+    case :ets.first(queue) do
+      {_time, holder} = key ->
+        checkout_holder(holder, from, queue) and :ets.delete(queue, key)
+        {:noreply, ready}
+      :"$end_of_table" ->
+        handle_info(checkout, put_elem(ready, 0, :busy))
```

Comment:
I think it would be clearer if we had `handle_info(checkout, {:busy, queue, codel})` here. I think the `put_elem` call is a bit distracting

Reply:
put_elem is faster, we could have helpers for updating the state and inline them.

Reply:
I don't think the perf difference will be noticeable here 😛

Reply:
Will refactor this part for clarity and avoid some pattern matching.